src/main/java/ch/njol/skript/expressions/ExprLastAttack.java

```diff
@@ -0,0 +1,152 @@
+		setExpr((Expression<? extends Entity>) exprs[0]);
+		finalDamage = parseResult.hasTag("final");
+		attacker = parseResult.hasTag("attacker");
+		cause = parseResult.hasTag("cause");
+		return true;
+	}
+
+	@Override
+	@Nullable
+	public Object convert(Entity entity) {
+		if (attacker)
+			return ExprAttacker.getAttacker(entity.getLastDamageCause());
+		if (finalDamage)
+			return entity.getLastDamageCause().getFinalDamage();
+		if (cause)
+			return entity.getLastDamageCause().getCause();
+		return entity.getLastDamageCause().getDamage();
+	}
+
+	@Override
+	@Nullable
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		if (!SETTABLE || mode != ChangeMode.SET)
+			return null;
+		if (attacker)
+			return CollectionUtils.array(Entity.class);
+		if (cause)
+			return CollectionUtils.array(DamageCause.class);
+		return CollectionUtils.array(Number.class);
+	}
+
+	@Override
+	@SuppressWarnings("deprecation")
+	public void change(Event event, @Nullable Object[] delta, ChangeMode mode) {
+		if (delta == null)
+			return;
+		if (attacker) {
+			Entity damager = (Entity) delta[0];
+			for (Entity entity : getExpr().getArray(event)) {
+				EntityDamageEvent damageEvent = entity.getLastDamageCause();
+				// Entity has not been harmed.
+				if (damageEvent == null) {
+					entity.setLastDamageCause(new EntityDamageByEntityEvent(damager, entity, DamageCause.CUSTOM, 0D));
+					continue;
+				}
+				entity.setLastDamageCause(new EntityDamageByEntityEvent(damager, entity, damageEvent.getCause(), damageEvent.getDamage()));
+			}
+		} else if (cause) {
+			DamageCause cause = (DamageCause) delta[0];
+			for (Entity entity : getExpr().getArray(event)) {
+				EntityDamageEvent damageEvent = entity.getLastDamageCause();
+				// We have no entity, the entity cannot be null in the EntityDamageEvent constructor.
+				if (damageEvent == null)
+					continue;
+				// Future proof
+				if (damageEvent instanceof EntityDamageByBlockEvent) {
+					entity.setLastDamageCause(new EntityDamageByBlockEvent(((EntityDamageByBlockEvent)damageEvent).getDamager(), entity, cause, damageEvent.getDamage()));
+					continue;
+				}
+				entity.setLastDamageCause(new EntityDamageByEntityEvent(damageEvent.getEntity(), entity, cause, damageEvent.getDamage()));
+			}
+		} else {
+			double damage = ((Number) delta[0]).doubleValue();
+			for (Entity entity : getExpr().getArray(event)) {
+				EntityDamageEvent damageEvent = entity.getLastDamageCause();
+				// We have no entity, the entity cannot be null in the EntityDamageEvent constructor.
+				if (damageEvent == null)
+					continue;
+				// Future proof
+				if (damageEvent instanceof EntityDamageByBlockEvent) {
+					entity.setLastDamageCause(new EntityDamageByBlockEvent(((EntityDamageByBlockEvent)damageEvent).getDamager(), entity, damageEvent.getCause(), damage));
+					continue;
+				}
+				entity.setLastDamageCause(new EntityDamageByEntityEvent(damageEvent.getEntity(), entity, damageEvent.getCause(), damage));
```

Comment:
This line is slightly wrong - there is an EntityDamageEvent simpliciter that isn't abstract, which returns `getEntity` as the victim, which, hopefully, is the same as our `entity`.
However, EntityDamageByEntityEvent uses the damager as its first parameter and passes the _victim_ up to EntityDamageEvent:
<img width="667" alt="image" src="https://user-images.githubusercontent.com/14147477/227981977-38b0455c-d079-4059-aef7-568d4bc86518.png">
The damager is available through `getDamager`
<img width="514" alt="image" src="https://user-images.githubusercontent.com/14147477/227982069-a5a8d5fb-94d4-4870-b7ac-f25e196ef383.png">

So we should instead be checking if the event is an instance of `EntityDamageByEntityEvent` and then pass the correct damager and victim to the new constructor, and use an `EntityDamageEvent` as our fall-back, otherwise we will change the event to be wrong.

Reply:
In fact, all of the setters may need changing.